official/recommendation/neumf_model.py

```diff
@@ -0,0 +1,182 @@
+# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Defines NeuMF model for NCF framework.
+
+Some abbreviations used in the code base:
+NeuMF: Neural Matrix Factorization
+NCF: Neural Collaborative Filtering
+GMF: Generalized Matrix Factorization
+MLP: Multi-Layer Perceptron
+
+GMF applies a linear kernel to model the latent feature interactions, and MLP
+uses a nonlinear kernel to learn the interaction function from data. NeuMF model
+is a fused model of GMF and MLP to better model the complex user-item
+interactions, and unifies the strengths of linearity of MF and non-linearity of
+MLP for modeling the user-item latent structures.
+
+In NeuMF model, it allows GMF and MLP to learn separate embeddings, and combine
+the two models by concatenating their last hidden layer.
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf  # pylint: disable=g-bad-import-order
+
+from six.moves import xrange  # pylint: disable=redefined-builtin
+
+
+def _validate_batch_size_for_multi_gpu(batch_size):
+  """For multi-gpu, batch-size must be a multiple of the number of GPUs.
```

Comment:
Perhaps add an explanation about why the batch size needs to be a multiple (is the batch being split across each of the GPUs?)/

Reply:
Yes. The batch will be evenly split across all GPUs. Will add comments there.